crates/sui-core/src/congestion_tracker.rs

```diff
@@ -0,0 +1,331 @@
+            self.highest_cancelled_gas_price = new.highest_cancelled_gas_price;
+        }
+        // If there are more recent successful transactions, we need to know the latest lowest
+        // executed price
+        if new.last_success_time > self.last_success_time {
+            self.last_success_time = new.last_success_time;
+            self.lowest_executed_gas_price = new.lowest_executed_gas_price;
+        }
+    }
+
+    fn update_for_cancellation(&mut self, now: CheckpointTimestamp, gas_price: u64) {
+        self.last_cancellation_time = now;
+        self.highest_cancelled_gas_price =
+            std::cmp::max(self.highest_cancelled_gas_price, gas_price);
+    }
+
+    fn update_for_success(&mut self, now: CheckpointTimestamp, gas_price: u64) {
+        self.last_success_time = Some(now);
+        self.lowest_executed_gas_price = Some(match self.lowest_executed_gas_price {
+            Some(current_min) => std::cmp::min(current_min, gas_price),
+            None => gas_price,
+        });
+    }
+}
+
+pub struct CongestionTracker {
+    pub congestion_clearing_prices: Cache<ObjectID, CongestionInfo>,
+}
+
+impl Default for CongestionTracker {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl CongestionTracker {
+    pub fn new() -> Self {
+        Self {
+            congestion_clearing_prices: Cache::new(10_000),
+        }
+    }
+
+    pub fn process_checkpoint_effects(
+        &self,
+        transaction_cache_reader: &dyn TransactionCacheRead,
+        checkpoint: &VerifiedCheckpoint,
+        effects: &[TransactionEffects],
+    ) {
+        let mut congestion_events = Vec::with_capacity(effects.len());
+        let mut cleared_events = Vec::with_capacity(effects.len());
+
+        for effect in effects {
+            let gas_price = transaction_cache_reader
+                .get_transaction_block(effect.transaction_digest())
+                .unwrap()
+                .transaction_data()
+                .gas_price();
+            if let Some(CongestedObjects(congested_objects)) =
+                effect.status().get_congested_objects()
+            {
+                congestion_events.push((gas_price, congested_objects.clone()));
+            } else {
+                cleared_events.push((
+                    gas_price,
+                    effect
+                        .input_shared_objects()
+                        .into_iter()
+                        .filter_map(|object| match object {
+                            InputSharedObject::Mutate((id, _, _)) => Some(id),
+                            InputSharedObject::Cancelled(_, _)
+                            | InputSharedObject::ReadOnly(_)
+                            | InputSharedObject::ReadDeleted(_, _)
+                            | InputSharedObject::MutateDeleted(_, _) => None,
+                        })
+                        .collect::<Vec<_>>(),
+                ));
+            }
+        }
```

Comment:
I assume we only care if we  don't have congested objects, not just that the txn succeeded (which is why this logic doesn't actually look at the success status)

Reply:
congested objects indicates that the tx was aborted due to congestion.